Dashboard: GUI reflects DISABLE_SEARCH_INDEXING

jrwdunham opened the pull request:
If Archivematica has been deployed with Elasticsearch indexing disabled, then the following changes will result:

1. The Backlog, Appraisal, and Archival storage tabs will now no longer be displayed in the navigation bar. If the user navigates to them using their URLs, the pages will indicate their non-operationality.
2. The SIP Arrange pane of the Ingest tab will no longer be displayed.
3. The disabled status of ES indexing will be indicated under Administration > General.
4. The "Send to backlog" option of "Create SIP(s)" will not be available in the Processing Configuration edit pages.
5. When the "Create SIP(s)" choice is displayed to the user during transfer, it will no longer contain a "Send to backlog" option.

Fixes https://github.com/artefactual/archivematica/issues/803 and implements feature file at https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/50.

Possible TODO: If the user supplies a processingConfig.xml file in a transfer that causes "Create SIP(s)" to default to "Send to backlog", then MCPServer should silently ignore this.

sevein commented:
@jrwdunham we can't merge this into qa/1.x because AM17 hasn't been released yet. Should't you be targetting dev/1.8.x instead?

jrwdunham committed:
Dashboard: reflect DISABLE_SEARCH_INDEXING in GUI

If Archivematica has been deployed with Elasticsearch indexing disabled,
then the Backlog, Appraisal, and Archival storage tabs will now no
longer be displayed. In addition, the SIP Arrange pane in the Ingest tab
is no longer displayed. Finally, the status of ES indexing is also indicated
under Administration > General.

jrwdunham committed:
Dashboard: fix processing config for indexless

If ES indexing is turned off (i.e., setting `DISABLE_SEARCH_INDEXING` is set
to `True`), then the processing config edit page will no longer display
the "Send to backlog" option at the "Create SIP(s)" choice.

jrwdunham committed:
MCPServer: fix choices for indexless

If ES indexing is disabled, then the MCPServer will no longer cause the
"Send to backlong" option to be available at the "Create SIP(s)" choice
point.

sevein commented on src/MCPServer/lib/linkTaskManagerChoice.py:
```diff
@@ -57,6 +57,9 @@ def __init__(self, jobChainLink, pk, unit):
         choices = MicroServiceChainChoice.objects.filter(choiceavailableatlink_id=str(jobChainLink.pk))
         for choice in choices:
+            if (django_settings.DISABLE_SEARCH_INDEXING and
+                    choice.chainavailable.description == 'Send to backlog'):
```
Could we instead have a number of abilities and hide these items based on that? E.g. pipeline `X` has the `search` ability disabled, choice `Y` depends on `search` ability so it's ignored.

jrwdunham replied:
Yes, I think that's a good idea. I've tried to implement what you've described in the most recent commit. Let me know what you think.

sevein commented on src/dashboard/src/templates/appraisal/appraisal.html:
```diff
@@ -20,9 +20,14 @@
+  {% else %}
+    {% verbatim %}
+    <div ng-app="appraisalTab">
+          <ng-include src="'front_page/content.html'"></ng-include>
```
Fix indentation.

jrwdunham replied:
fixed

sevein commented on src/dashboard/src/templates/archival_storage/list.html:
```diff
@@ -25,103 +25,110 @@
-  </ul>
+  {% if disable_search_indexing %}
+    <h1>Elasticsearch Indexing Disabled</h1>
+    <p class="es-indexing-disabled">Sorry, Elasticsearch indexing has been disabled in this Archivematica installation. The archival storage tab is non-functional when indexing is turned off.</p>
```
Fix i18n.

jrwdunham replied:
Fixed here and also in all newly added template strings.

sevein commented:
@jrwdunham, I've rebased `dev/1.8.x` because it was behind qa/1.x missing some important fixes.

I've also created `stable/1.7.x` so you can start submitting pull requests against `qa/1.x` for AM18 stuff. Perhaps you can repurpose this PR to include what's in `dev/1.8.x` plus changes in `dev/issue-11442-indexless-GUI`. We should delete `dev/1.8.x`, `dev/1.8.x-indexless-GUI` and `dev/1.8.x-without-logs` I assume?

jrwdunham commented:
@sevein I've rebased this against the new dev/1.8.x. Can you CR again?

The dev/1.8.x-indexless-GUI and dev/1.8.x-without-logs branches are gone.

sevein reviewed (changes requested):
Two observations:

- I think that `Add common abilities with choice dependencies` should be updated
so it doesn't include the setting rename because it's unrelated.

- Regarding to `Dashboard: GUI reflects DISABLE_SEARCH_INDEXING`, I think as a
developer my only setting interface should be `django.conf.settings`. I don't
want to remember what `get_search_enabled` is, I want to simply look up
`django_settings.SEARCH_ENABLED`. I think your motivation is good because the
existing name sucks but I'm not confortable solving that problem this way
because it's not cohesive. IMO it would be a win win (for users and for us) to
implement it in `appconfig.Config` and add some tests - and if you don't have
time for that then I would just ignore this for now. Can we have something like
this in `CONFIG_MAPPING`?

```
'search_enabled': [
    {'section': 'MCPClient', 'option': 'disableElasticsearchIndexing', 'type': 'boolean'},
    {'section': 'MCPClient', 'option': 'search_enabled', 'type': 'iboolean'},
}
```

Two changes would be required:

- `appconfig.Config.get` would need to handle lists.
- `env_configparser.EnvConfigParser` would need to have a new `iboolean` method
added (you could subclass `EnvConfigParser` too).

jrwdunham committed:
Add common abilities with choice dependencies

New abilities.py module in archivematicaCommon defines abilities of an
Archivematica instance as a whole. An ability can have dependencies. For
example, the Backlog feature is dependent on the Search ability. If the Search
(i.e., Elasticsearch indexing) ability is disabled, then the Backlog tab should
be unavailable; in addition, the choice to send a transfer to backlog (either
via pre-configuration or via real-time choice in the GUI) should not be
available either.

Some minor formatting and i18n changes were also made.

jrwdunham committed:
Enhance appconfig to allow multiple equivalent options

The `appconfig.py::Config` class accepts an `attrs` dict argument whose values
can now contain lists of dicts as opposed to just single dicts as previsouly.
This allows one to specify multiple configuration options as equivalent and
reduce them to a single option. Using this functionality, the
`disable_search_indexing` and `search_enabled` config options are declared to
be equivalent, the one being the inverse of the other. The resulting settings
attribute is accessed as `settings.SEARCH_ENABLED`.

sevein commented on src/MCPClient/lib/settings/common.py:
```diff
@@ -58,6 +61,8 @@
     'db_port': {'section': 'client', 'option': 'port', 'type': 'string'},
 }
 
+SEARCH_ENABLED_DEFAULT = False
```
Do we still need this?

jrwdunham replied:
No, I guess not

sevein commented on src/archivematicaCommon/lib/env_configparser.py:
```diff
@@ -70,3 +70,7 @@ def getfloat(self, *args, **kwargs):
+
+    @fallback_option
+    def getiboolean(self, *args, **kwargs):
+        return not self.getboolean(*args, **kwargs)
```
`tests/test_env_configparser.py` has already good coverage, should we include a test for this new method?

jrwdunham replied:
I've added new tests for appconfig that test the new "multiple equivalent options" feature.

jrwdunham committed:
archivematicaCommon: add tests for appconfig

jrwdunham committed:
Remove SEARCH_ENABLED_DEFAULT constant from settings
